src/main/java/com/google/jspecify/nullness/ConformanceTypeInformationPresenter.java

```diff
@@ -57,50 +60,56 @@ class ConformanceTypeInformationReporter extends TypeInformationReporter {
+        case ASSIGNMENT:
+          AssignmentTree asgn = (AssignmentTree) tree;
+          AnnotatedTypeMirror varType;
+          if (genFactory != null) {
+            varType = genFactory.getAnnotatedTypeLhs(asgn.getVariable());
+          } else {
+            varType = atypeFactory.getAnnotatedType(asgn.getVariable());
+          }
+          checker.reportWarning(
+              asgn.getVariable(),
+              "sinkType",
+              typeFormatter.format(varType),
+              asgn.getVariable().toString());
+          checker.reportWarning(
+              asgn.getExpression(),
+              "sourceType",
+              typeFormatter.format(atypeFactory.getAnnotatedType(asgn.getExpression())),
+              asgn.getExpression().toString());
+          break;
         case RETURN:
-          direction = "sinkType";
-          usage = "return";
+          checker.reportWarning(tree, "sinkType", typeFormatter.format(type), "return");
           break;
-
         case METHOD_INVOCATION:
+          ExecutableElement calledElem = TreeUtils.elementFromUse((MethodInvocationTree) tree);
+          String methodName = calledElem.getSimpleName().toString();
           AnnotatedExecutableType calledType = (AnnotatedExecutableType) type;
           List<? extends AnnotatedTypeMirror> params = calledType.getParameterTypes();
           MethodInvocationTree mit = (MethodInvocationTree) tree;
           List<? extends ExpressionTree> args = mit.getArguments();
           assert params.size() == args.size();
 
           for (int i = 0; i < params.size(); ++i) {
-            checker.reportError(
-                tree,
-                "sinkType",
-                typeFormatter.format(params.get(i)),
-                // TODO: build up something from method and parameter name?
-                "param");
-
-            checker.reportError(
+            String paramName = calledElem.getParameters().get(i).getSimpleName().toString();
+            String paramLocation = String.format("%s#%s", methodName, paramName);
+            checker.reportWarning(
+                tree, "sinkType", typeFormatter.format(params.get(i)), paramLocation);
+            checker.reportWarning(
                 tree,
                 "sourceType",
                 typeFormatter.format(atypeFactory.getAnnotatedType(args.get(i))),
-                // TODO: build up something from method and parameter name?
-                "arg");
-          }
-          break;
-        case IDENTIFIER:
-          if (occurrenceKind == TypeOccurrenceKind.USE_TYPE) {
-            direction = "sourceType";
-            usage = tree.toString();
+                args.get(i).toString());
           }
           break;
         default:
+          // Nothing special for other trees.
       }
 
-      if (direction != null && usage != null) {
-        checker.reportError(tree, direction, typeFormatter.format(type), usage);
+      if (TreeUtils.isExpressionTree(tree)) {
```

Comment:
Do you need the previous "sourceType" reports at all if you're doing this?

Reply:
Right! In an earlier version I output the argument position too, but that went away, so this seems to be enough for all `sourceType`s.